flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/TimeBoundedStreamJoinOperator.java

```diff
@@ -0,0 +1,480 @@
+	 * defined boundaries, it gets collected.
+	 *
+	 * @param record An incoming record to be joined
+	 * @throws Exception Can throw an exception during state access
+	 */
+	@Override
+	public void processElement2(StreamRecord<T2> record) throws Exception {
+
+		long rightTs = record.getTimestamp();
+		T2 rightElem = record.getValue();
+
+		long joinLowerBound = rightTs + inverseLowerBound;
+		long joinUpperBound = rightTs + inverseUpperBound;
+
+		addToRightBuffer(rightElem, rightTs);
+
+		if (rightTs == Long.MIN_VALUE) {
+			throw new RuntimeException("Time-bounded stream joins need to have timestamps " +
+				"assigned to elements, but current element has timestamp Long.MIN_VALUE");
+		}
+
+		if (dataIsLate(rightTs)) {
+			return;
+		}
+
+		for (Map.Entry<Long, List<Tuple3<T1, Long, Boolean>>> entry : leftBuffer.entries()) {
+			long bucketStart = entry.getKey();
+			long bucketEnd = bucketStart + bucketGranularity;
+
+			if (!(bucketEnd >= joinLowerBound && bucketStart <= joinUpperBound)) {
+				// skip buckets that are out of bounds
+				continue;
+			}
+
+			for (Tuple3<T1, Long, Boolean> tuple : entry.getValue()) {
+				if (shouldBeJoined(tuple.f1, rightTs)) {
+
+					// collect joined tuple with left timestamp
+					collect(tuple.f0, rightElem, tuple.f1, rightTs);
+				}
+			}
+		}
+
+		registerCleanupTimer();
+	}
+
+	private boolean dataIsLate(long rightTs) {
+		return this.lastWatermark != Long.MIN_VALUE && rightTs < lastWatermark - watermarkDelay;
+	}
+
+	@Override
+	public void processWatermark(Watermark mark) throws Exception {
+
+		// We can not clean our state here directly because we are not in a keyed context. Instead
+		// we set a field containing the last watermark that we have seen, and for every element in
+		// processElement1(...) / processElement2(...) we register a timer with time: watermark + 1
+		// This watermark + 1 will then trigger the onEventTime(...) method for the next watermark,
+		// where we are in a keyed context again, which we can use to clean up our state.
+		this.lastWatermark = mark.getTimestamp();
+
+		if (timeServiceManager != null) {
+			timeServiceManager.advanceWatermark(mark);
+		}
+
+		// emit the watermark with the calculated delay, so we don't produce late data
+		output.emitWatermark(new Watermark(this.lastWatermark - watermarkDelay));
+	}
+
+	private void collect(T1 left, T2 right, long leftTs, long rightTs) throws Exception {
+		collector.setAbsoluteTimestamp(leftTs);
+		context.leftTs = leftTs;
+		context.rightTs = rightTs;
+		userFunction.processElement(left, right, context, this.collector);
+	}
+
+	private void removeFromLhsUntil(long maxCleanup) throws Exception {
+
+		Iterator<Map.Entry<Long, List<Tuple3<T1, Long, Boolean>>>> iterator = leftBuffer.iterator();
+		while (iterator.hasNext()) {
+			Map.Entry<Long, List<Tuple3<T1, Long, Boolean>>> next = iterator.next();
+			if (next.getKey() + bucketGranularity <= maxCleanup) {
+				iterator.remove();
```

Comment:
I'm not a big fan of the troublesome `iterator.remove()`. Will there be concurrent access to `left/rightBuffer` while removing elements from them?

Reply:
With the effectively single-threaded execution model of the operators I don't think there should be any problems here, as far as I understand. The only other thread accessing the state should be the checkpointing one, and I'd assume that this happens in a safe manner. 
I had a quick chat with @kl0u and he confirmed this